pkg/note/note.go

```diff
@@ -0,0 +1,152 @@
+	return n.sign(msg)
+}
+
+// isValidName reports whether the name conforms to the spec for the origin string of the note text
+// as defined in https://github.com/C2SP/C2SP/blob/main/tlog-checkpoint.md#note-text.
+func isValidName(name string) bool {
+	return name != "" && utf8.ValidString(name) && strings.IndexFunc(name, unicode.IsSpace) < 0 && !strings.Contains(name, "+")
+}
+
+// genConformantKeyHash generates a 4-byte key ID for typical (non-ECDSA) keys.
+func genConformantKeyHash(name string, sigType, key []byte) uint32 {
+	hash := sha256.New()
+	hash.Write([]byte(name))
+	hash.Write([]byte("\n"))
+	hash.Write(sigType)
+	hash.Write(key)
+	sum := hash.Sum(nil)
+	return binary.BigEndian.Uint32(sum)
+}
+
+// ed25519KeyHash generates the 4-byte key ID for an Ed25519 public key.
+// Ed25519 keys are the only key type compatible with witnessing.
+func ed25519KeyHash(name string, key []byte) uint32 {
+	return genConformantKeyHash(name, []byte{algEd25519}, key)
+}
+
+// ecdsaKeyHash generates the 4-byte key ID for an ECDSA public key.
+// ECDSA key IDs do not conform to the note standard for other key type
+// (see https://github.com/C2SP/C2SP/blob/8991f70ddf8a11de3a68d5a081e7be27e59d87c8/signed-note.md#signature-types).
+func ecdsaKeyHash(key *ecdsa.PublicKey) (uint32, error) {
+	marshaled, err := x509.MarshalPKIXPublicKey(key)
+	if err != nil {
+		return 0, fmt.Errorf("marshaling public key: %w", err)
+	}
+	hash := sha256.Sum256(marshaled)
+	return binary.BigEndian.Uint32(hash[:]), nil
+}
+
+// rsaKeyhash generates the 4-byte key ID for an RSA public key.
+func rsaKeyHash(name string, key *rsa.PublicKey) (uint32, error) {
+	marshaled, err := x509.MarshalPKIXPublicKey(key)
+	if err != nil {
+		return 0, fmt.Errorf("marshaling public key: %w", err)
+	}
+	rsaAlg := append([]byte{algUndef}, []byte(rsaID)...)
+	return genConformantKeyHash(name, rsaAlg, marshaled), nil
+}
+
+// NewNoteSigner converts a sigstore/sigstore/pkg/signature.Signer into a note.Signer.
+func NewNoteSigner(ctx context.Context, origin string, signer signature.Signer) (note.Signer, error) {
+	if !isValidName(origin) {
+		return &noteSigner{}, fmt.Errorf("invalid name %s", origin)
+	}
+
+	pubKey, err := signer.PublicKey()
+	if err != nil {
+		return &noteSigner{}, fmt.Errorf("getting public key: %w", err)
+	}
+	var keyID uint32
+	switch pk := pubKey.(type) {
+	case *ecdsa.PublicKey:
+		keyID, err = ecdsaKeyHash(pk)
+		if err != nil {
+			return &noteSigner{}, fmt.Errorf("getting ECDSA key hash: %w", err)
+		}
+	case ed25519.PublicKey:
+		keyID = ed25519KeyHash(origin, pk)
+	case *rsa.PublicKey:
+		keyID, err = rsaKeyHash(origin, pk)
+		if err != nil {
+			return &noteSigner{}, fmt.Errorf("getting RSA key hash: %w", err)
+		}
+	default:
+		return &noteSigner{}, fmt.Errorf("unsupported key type: %T", pubKey)
+	}
+
+	sign := func(msg []byte) ([]byte, error) {
+		return signer.SignMessage(bytes.NewReader(msg), options.WithContext(ctx))
```

Comment:
Could context be an input into the SignMessage function rather than the NewNoteSigner func?

Reply:
It's not possible without breaking the note signer interface. I wouldn't expect the program to want to create a new context every time it signs something, it would typically use a long running server context.

Reply:
Oh I was thinking that each signing event would be cancellable/timeoutable and that's why it would have a separate context config

Reply:
If this was tied to requests then that would be desirable, but checkpoint signing happens in batches and isn't 1:1 with incoming requests. If I'm wrong and we do want the signing event to be cancellable independent of the main server context then we could always have multiple instances of the signer.

Reply:
Ah I see, okay cool thanks for the clarification.